src/lifetimes.rs

```diff
@@ -159,35 +157,75 @@ fn unique_lifetimes(lts: &[RefLt]) -> usize {
 }
 
 /// A visitor usable for rustc_front::visit::walk_ty().
-struct RefVisitor(Vec<RefLt>);
+struct RefVisitor<'v, 't: 'v> {
+    cx: &'v LateContext<'v, 't>, // context reference
+    lts: Vec<RefLt>
+}
+
+impl <'v, 't> RefVisitor<'v, 't>  {
+    fn new(cx: &'v LateContext<'v, 't>) -> RefVisitor<'v, 't> {
+        RefVisitor { cx: cx, lts: Vec::new() }
+    }
 
-impl RefVisitor {
     fn record(&mut self, lifetime: &Option<Lifetime>) {
         if let &Some(ref lt) = lifetime {
             if lt.name.as_str() == "'static" {
-                self.0.push(Static);
+                self.lts.push(Static);
             } else {
-                self.0.push(Named(lt.name));
+                self.lts.push(Named(lt.name));
             }
         } else {
-            self.0.push(Unnamed);
+            self.lts.push(Unnamed);
         }
     }
 
     fn into_vec(self) -> Vec<RefLt> {
-        self.0
+        self.lts
+    }
+
+    fn collect_anonymous_lifetimes(&mut self, path: &Path, ty: &Ty) {
+        let last_path_segment = path.segments.last().map(|s| &s.parameters);
+        if let Some(&AngleBracketedParameters(ref params)) = last_path_segment {
+            if params.lifetimes.is_empty() {
+                let def = self.cx.tcx.def_map.borrow().get(&ty.id).map(|r| r.full_def());
+                match def {
+                    Some(DefTy(def_id, _)) => {
+                        if let Some(ty_def) = self.cx.tcx.adt_defs.borrow().get(&def_id) {
+                            let scheme = ty_def.type_scheme(self.cx.tcx);
+                            for _ in scheme.generics.regions.as_slice() {
+                                self.record(&None);
+                            }
+                        }
+                    },
+                    Some(DefTrait(def_id)) => {
+                        let trait_def = self.cx.tcx.trait_defs.borrow()[&def_id];
+                        for _ in &trait_def.generics.regions {
+                            self.record(&None);
+                        }
+                    },
+                    _ => {}
+                }
+            }
+        }
     }
 }
 
-impl<'v> Visitor<'v> for RefVisitor {
+impl<'v, 't> Visitor<'v> for RefVisitor<'v, 't> {
+
     // for lifetimes as parameters of generics
     fn visit_lifetime(&mut self, lifetime: &'v Lifetime) {
         self.record(&Some(*lifetime));
     }
 
     fn visit_ty(&mut self, ty: &'v Ty) {
-        if let TyRptr(None, _) = ty.node {
-            self.record(&None);
+        match ty.node {
+            TyRptr(None, _) => {
+                self.record(&None);
+            },
+            TyPath(_, ref path) => {
+                self.collect_anonymous_lifetimes(path, ty);
+            },
+            _ => {}
```

Comment:
What is the problem with those? I've tested this on functions with trait object arguments, and I haven't noticed any problems.